zkevm-circuits/src/ecc_circuit.rs

```diff
@@ -0,0 +1,741 @@
+//! The ECC circuit is responsible for verifying ECC-related operations from precompiled contract
+//! calls, namely, EcAdd, EcMul and EcPairing.
+
+use std::marker::PhantomData;
+
+use bus_mapping::{
+    circuit_input_builder::{EcAddOp, EcMulOp, EcPairingOp},
+    precompile::PrecompileCalls,
+};
+use eth_types::{Field, ToScalar};
+use halo2_base::{
+    gates::{GateInstructions, RangeInstructions},
+    utils::modulus,
+    Context, QuantumCell, SKIP_FIRST_PASS,
+};
+use halo2_ecc::{
+    bn254::pairing::PairingChip,
+    ecc::EccChip,
+    fields::{
+        fp::{FpConfig, FpStrategy},
+        fp12::Fp12Chip,
+        FieldChip,
+    },
+};
+use halo2_proofs::{
+    arithmetic::Field as Halo2Field,
+    circuit::{Layouter, Value},
+    halo2curves::bn256::{Fq, Fq12, Fr, G1Affine, G2Affine},
+    plonk::{ConstraintSystem, Error, Expression},
+};
+use itertools::Itertools;
+use log::error;
+
+use crate::{
+    table::{EccTable, LookupTable},
+    util::{Challenges, SubCircuit, SubCircuitConfig},
+    witness::Block,
+};
+
+mod dev;
+mod test;
+mod util;
+use util::{
+    EcAddAssigned, EcMulAssigned, EcOpsAssigned, EcPairingAssigned, G1Assigned, G1Decomposed,
+    G2Assigned, G2Decomposed, ScalarAssigned, ScalarDecomposed,
+};
+
+use self::util::LOG_TOTAL_NUM_ROWS;
+
+/// Arguments accepted to configure the EccCircuitConfig.
+#[derive(Clone, Debug)]
+pub struct EccCircuitConfigArgs<F: Field> {
+    /// ECC table that is connected to the ECC circuit.
+    pub ecc_table: EccTable,
+    /// zkEVM challenge API.
+    pub challenges: Challenges<Expression<F>>,
+}
+
+/// Config for the ECC circuit.
+#[derive(Clone, Debug)]
+pub struct EccCircuitConfig<F: Field> {
+    /// Field config for halo2_proofs::halo2curves::bn256::Fq.
+    fp_config: FpConfig<F, Fq>,
+    /// Lookup table for I/Os to the EcAdd, EcMul and EcPairing operations.
+    ecc_table: EccTable,
+
+    /// Number of limbs to represent Fp.
+    num_limbs: usize,
+    /// Number of bits per limb.
+    limb_bits: usize,
+
+    _marker: PhantomData<F>,
+}
+
+impl<F: Field> SubCircuitConfig<F> for EccCircuitConfig<F> {
+    type ConfigArgs = EccCircuitConfigArgs<F>;
+
+    fn new(
+        meta: &mut ConstraintSystem<F>,
+        Self::ConfigArgs {
+            ecc_table,
+            challenges: _,
+        }: Self::ConfigArgs,
+    ) -> Self {
+        let num_limbs = 3;
+        let limb_bits = 88;
+        let fp_config = FpConfig::configure(
+            meta,
+            FpStrategy::Simple,
+            &[30], // num advice
+            &[17], // num lookup advice
+            1,     // num fixed
+            13,    // lookup bits
```

Comment:
Why is lookup bits set to 13? The sig circuit uses 17 there.

Reply:
Actually I used the same config parameters as in the sig circuit.